fix(color): Logical switches monitor fix

banyaszg opened the pull request:
Fixes #2384

Summary of changes:
This patch uses styles for active and non active buttons. The style defines colors and fonts for each logical switch state.

raphaelcoeffic commented:
I would have strongly preferred the ButtonMatrix. Was there an issue with taking that route?

we should avoid having Styles created and applied in object constructors when possible, as it is rather slow. It this particular case it is also leaking memory, as the styles need to be destroyed after use (LVGL is not C++).

banyaszg commented:
I implemented with ButtonMatrix first. It worked well except one behaviour. The focus/editing thing caused a lot of problem. The whole matrix was a widget, so normaly you need to push the Enter to start editing mode. Then you can scroll the focus with the scroll button, but if you press the Enter again, then you leave the editing mode and lose the selection. I spent hours to find an elegant solution, but it required so much hacking I gave up around 1am. Then I tried to fix the current code, and this way looked good. This way it works, but the implementation is not valid yet, so I thought I will sleep first and refine it after.

banyaszg commented:
Actually we can do with direct change too skipping the styles, just I thought it is a good was to create a complete setting for each state.

banyaszg commented:
@raphaelcoeffic What do you think about defining and using a custom state as it is in this version?

raphaelcoeffic commented:
> What do you think about defining and using a custom state as it is in this version?

This is the right way to go, for sure. However, this is basically the same as `LV_STATE_CHECKED`, so that you should not have to use a custom state for that.

The only difference I see here to the style that covers that state is:
- Text color `SECONDARY1` instead of `PRIMARY2`
- **Bold** font

I understand the use of **Bold**, but we should probably keep the same colours for sake of consistency.

banyaszg commented:
The style settings are from the orignal code to look the same as it was.
I saw the similarities with the `LV_STATE_CHECKED` style, but I thought the color differs too much. Actually the `PRIMARY2` is white in the default theme and the white text on yellow background moderately readable. I think it is ok for one or two options, because I know what that thing is, but in middle of such a big matrix it counts.

raphaelcoeffic commented:
> Actually the PRIMARY2 is white in the default theme and the white text on yellow background moderately readable.

This is something lots of people have complained about. So maybe we should change the generic style to use `SECONDARY1` instead when background is `ACTIVE`? That would leave us with just the **bold** font here.

banyaszg commented:
> This is something lots of people have complained about. So maybe we should change the generic style to use `SECONDARY1` instead when background is `ACTIVE`? That would leave us with just the **bold** font here.

I think it is a good idea to change that color. After a certain age we really appreciate the contrast. Also the **bold** font could be default for the checked state. However the `PRIMARY1` color can be enough without **bold**. 
And with the proper contrast the `LV_STATE_CHECKED` would be good in this part.

pfeerick commented:
> After a certain age we really appreciate the contrast.

Age has nothing to do with it... I'm not *that* old and I still appreciate good contrast... 😀

banyaszg commented:
> Age has nothing to do with it... I'm not _that_ old and I still appreciate good contrast... grinning

I'm also not old, just my minimum focus distance is increasing.
But at 18 meters I still hit the 10. :grinning:

gagarinlg commented:
> > Age has nothing to do with it... I'm not _that_ old and I still appreciate good contrast... grinning
> 
> I'm also not old, just my minimum focus distance is increasing.

I am slowly getting in the area where things are either too far away or too close, but there is no point, where I can look sharp.

pfeerick commented:
Thankfully (or not) mine should never change now... had cataract surgery a couple of years ago so fixed focus vision now - unless the warranty runs out on the lenses 🙃

banyaszg commented:
I assume the default `LV_STATE_CHECKED` style will be changed, so I modified the code to use it.
Also I modified the layout for the portrait mode to look better. (Tested in NV14 simulator.)

pfeerick commented:
This generally looks a lot better, as well as on the NV14. Thanks for that!
